### ISSUE TITLE
OnServerLeave triggered at incorrect times

### ISSUE THREAD

**Issue by Davinatoratoe** [Pryaxis/TShock#1490]
## When does it get triggered?
* When the server first starts.
* When a player connects (in-between connecting and joining)
* When the server exits (using /exit in console)

## Code
* Uses the latest references for the latest version of TShock.
```
using System;
using TShockAPI;
using Terraria;
using TerrariaApi.Server;

namespace LeaveHookTest
{
    [ApiVersion(2, 1)]
    public class LeaveHookTest : TerrariaPlugin
    {
        public override string Name { get { return "LeaveHookTest"; } }
        public override string Author { get { return "MrGgggg243"; } }
        public override string Description { get { return "LeaveHookTest"; } }
        public override Version Version { get { return new Version(1, 0); } }

        public LeaveHookTest(Main game) : base(game)
        {
            Order = +4;
        }

        public override void Initialize()
        {
            ServerApi.Hooks.ServerLeave.Register(this, OnServerLeave);
        }

        void OnServerLeave(LeaveEventArgs args)
        {
            //Writing: if (TShock.IsShuttingDown) return; prevents the Unhandled exception when the server shuts down.

            TShock.Log.Write("-> OnServerLeave()", System.Diagnostics.TraceLevel.Info);
            TShock.Log.Write("-> " + args.Who + " left the server", System.Diagnostics.TraceLevel.Info);
        }
    }
}
```

## Console
* I started the server, joined the server, left the server, and exited the server. Here is the console by the end:
```
Terraria Server v1.3.5.3

Listening on port 7777
Type 'help' for a list of commands.

: 127.0.0.1:62063 is connecting...
MrSanta243 authenticated successfully as user MrSanta243.
MrSanta243 has joined.
MrSanta243 has left.
/exit
Server executed: /exit.
Saving world. Momentary lag might result from this.
Saving world data: 4%
Saving world data: 10%
Saving world data: 15%
Saving world data: 20%
Saving world data: 25%
Saving world data: 30%
Saving world data: 36%
Saving world data: 43%
Saving world data: 49%
Saving world data: 55%
Saving world data: 61%
Saving world data: 68%
Saving world data: 74%
Saving world data: 79%
Saving world data: 86%
Saving world data: 92%
Saving world data: 98%
Saving world data: 100%
Validating world save: 39%
Validating world save: 98%
World saved.
Server shutting down!
Unhandled exception
System.UnhandledExceptionEventArgs

Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at TerrariaApi.Server.LogWriterManager.ServerWriteLine(String message, TraceLevel kind)
   at TerrariaApi.Server.HandlerCollection`1.Invoke(ArgsType args)
   at TerrariaApi.Server.HookManager.InvokeServerLeave(Int32 who)
   at TerrariaApi.Server.Hooking.ServerHooks.OnPreReset(RemoteClient remoteClient)
   at OTAPI.Callbacks.Terraria.RemoteClient.PreReset(RemoteClient remoteClient)
   at Terraria.RemoteClient.Reset()
   at Terraria.Netplay.ServerLoop(Object threadContext)
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)
```

## Log
* And here is the server log
```
2017-09-08 08:33:37 - TextLog: INFO: TShock 4.3.24.0 (Mintaka) now running.
2017-09-08 08:33:37 - TShock: INFO: AutoSave Enabled
2017-09-08 08:33:37 - TShock: INFO: Backups Disabled
2017-09-08 08:33:37 - TShock: INFO: Welcome to TShock for Terraria. Initialization complete.
2017-09-08 08:33:47 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:33:47 - HandlerCollection`1: INFO: -> 0 left the server
2017-09-08 08:33:48 - GetDataHandlers: INFO: MrSanta243 authenticated successfully as user MrSanta243.
2017-09-08 08:33:48 - TShock: INFO: MrSanta243 (127.0.0.1) from 'superadmin' group joined. (1/8)
2017-09-08 08:33:48 - Utils: INFO: Broadcast: MrSanta243 has joined.
2017-09-08 08:33:52 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:33:52 - HandlerCollection`1: INFO: -> 0 left the server
2017-09-08 08:33:52 - Utils: INFO: Broadcast: MrSanta243 has left.
2017-09-08 08:33:52 - TShock: INFO: MrSanta243 disconnected.
2017-09-08 08:34:00 - Utils: INFO: Server executed: /exit.
2017-09-08 08:34:00 - Utils: INFO: Broadcast: Saving world. Momentary lag might result from this.
2017-09-08 08:34:01 - Utils: INFO: Broadcast: World saved.
2017-09-08 08:34:01 - SaveManager: INFO: World saved at (C:\Users\unknow\Documents\My Games\Terraria\Worlds\Creative_World.wld)
2017-09-08 08:34:01 - Utils: INFO: Broadcast: Server shutting down!
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 0 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 1 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 2 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 3 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 4 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 5 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 6 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 7 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 8 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 9 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 10 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 11 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 12 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 13 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 14 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 15 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 16 left the server
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> OnServerLeave()
2017-09-08 08:34:01 - HandlerCollection`1: INFO: -> 17 left the server
```

* When the server exits, the hook is triggered 18 times with args.Who as 0 - 17. The max player slots is only 8.

**Comment by ProfessorXZ** [comment]
I've tested this myself and got the same results. I couldn't reproduce the exception that occurs on exit, though. 

`OnServerLeave` is implemented through OTAPI's `˙RemoteClient.Reset` modification. The said modification provides a `PreReset` hook which TSAPI uses to invoke ServerLeave. This means that the hook is called each time a `RemoteClient` is reset, and this happens when a connection is accepted (which is why it fired upon connecting to the server) and when it is pending termination.

I haven't found a reason why it fires x times during exit, though.

**Comment by DeathCradle** [comment]
I was able to reproduce very similar results here, but my crash was due to SEC (i have the debug version).

![image](https://user-images.githubusercontent.com/776327/31699362-f3398e42-b405-11e7-8242-4da8845f6eeb.png)

Without SEC I cannot reproduce the crash. However, on topic I believe we should be wrapping the call to [`_hookManager.InvokeServerLeave(remoteClient.Id);` in a `remoteClient.IsActive` check.](https://github.com/Pryaxis/TerrariaAPI-Server/blob/bc9f7c69becb60816fc1e4587bbca15991231317/TerrariaServerAPI/TerrariaApi.Server/Hooking/ServerHooks.cs#L34)

**Comment by ivanbiljan** [comment]
Fixed in Pryaxis/TerrariaAPI-Server#146
